src/cli.rs

```diff
@@ -0,0 +1,273 @@
+            "json" => {
+                let print_schema = long("print-schema")
+                    .help("Print JSON schema and exit")
+                    .req_flag(());
+                construct!(CliArgs::Json { args, meta_args })
+                    .or_else(construct!(CliArgs::JsonSchema { print_schema }))
+            }
+            _ => unreachable!(),
+        };
+        let parser = Info::default().descr(descr_ext).for_parser(parser);
+        command(command_name, Some(descr), parser)
+    }
+
+    let publishers = subcommand_with_common_args(
+        "publishers",
+        args_parser.clone(),
+        metadata_args_parser.clone(),
+        "List all crates.io publishers in the depedency graph",
+        "Lists all crates.io publishers in the dependency graph and owned crates for each
+
+If a local cache created by 'update' subcommand is present and up to date,
+it will be used. Otherwise live data will be fetched from the crates.io API.",
+    );
+    let crates = subcommand_with_common_args(
+        "crates",
+        args_parser.clone(),
+        metadata_args_parser.clone(),
+        "List all crates in dependency graph and crates.io publishers for each",
+        "Lists all crates in dependency graph and crates.io publishers for each
+
+If a local cache created by 'update' subcommand is present and up to date,
+it will be used. Otherwise live data will be fetched from the crates.io API.",
+    );
+    let json = subcommand_with_common_args(
+        "json",
+        args_parser.clone(),
+        metadata_args_parser.clone(),
+        "Like 'crates', but in JSON and with more fields for each publisher",
+        "Detailed info on publishers of all crates in the dependency graph, in JSON
+
+The JSON schema is also available, use --print-schema to get it.
+
+If a local cache created by 'update' subcommand is present and up to date,
+it will be used. Otherwise live data will be fetched from the crates.io API.",
+    );
+
+    let cache_max_age = cache_max_age_parser.clone();
+    let update = construct!(CliArgs::Update { cache_max_age });
+    let update = Info::default()
+        .descr(
+            "Download the latest daily dump from crates.io to speed up other commands
+
+If the local cache is already younger than specified in '--cache-max-age' option,
+a newer version will not be downloaded.
+        
+Note that this downloads the entire crates.io database, which is hundreds of Mb of data!
+If you are on a metered connection, you should not be running the 'update' subcommand.
+Instead, rely on requests to the live API - they are slower, but use much less data.",
+        )
+        .for_parser(update);
+    let update = command(
+        "update",
+        Some("Download the latest daily dump from crates.io to speed up other commands"),
+        update,
+    );
+
+    let parser = publishers.or_else(crates).or_else(json).or_else(update);
+
+    Info::default()
```

Comment:
You might need to wrap it in a `supply-chain` command, I'm doing something similar here https://github.com/pacak/cargo-show-asm/blob/master/src/opts.rs#L192

Cargo passes all of the command line inside a child process and this includes `supply-chain` as a first argument in addition to executable name. Unless I'm using it wrong that is.

Reply:
Excellent catch! I completely forgot about that quirk of Cargo. Thank you!

Reply:
What I would like to accomplish is for `cargo supply-chain` have the help text as if the `supply-chain` argument doesn't exist - only show regular subcommands. And I also want it to work both with and without that argument being present.

With `pico-args` this was easy - just peek the `env::os_args()` iterator and skip the first argument if it is set to "supply-chain". I think I can potentially accomplish this with `bpaf` using `run_inner()` as well.

I would have to migrate my unit tests to a custom function, too; no more constructing `Args` directly, because I need my own shim there. But I can probably make that ergonomic using generics.

Reply:
Just wrapping it in a command should be enough. When you use `--help` on a subcommand - it displays only help for that subcommand and nothing else. For example my cargo-show-asm displays this, note total lack of `asm`.

For testing you can either pass `"supply-chain"` all over the place or do your tests on internal parser and only apply external wrapper when actually parsing the command line.

```
% cargo asm --help 
Usage: [--target-dir DIR] [--manifest-path PATH] [(--lib | --bin BIN | --test TEST | --bench BENCH | --example EXAMPLE)] -v... [-d] [-p SPEC] [--frozen] [--locked] [--offline] [-r] [--no-color] [<FUNCTION>]

Available options:
        --target-dir <DIR>      Custom target directory for generated artifacts
        --manifest-path <PATH>  Path to Cargo.toml
        --lib                   Show results from library code
        --bin <BIN>             Show results from a binary
        --test <TEST>           Show results from a test
        --bench <BENCH>         Show results from a benchmark
        --example <EXAMPLE>     Show results from an example
    -v, --verbose               more verbose output, can be specified multuple times
    -d, --dry                   Produce a build plan instead of actually building
    -p, --package <SPEC>        Package to use, if not specified
        --frozen                Require Cargo.lock and cache are up to date
        --locked                Require Cargo.lock is up to date
        --offline               Run without accessing the network
    -r, --rust                  Print interleaved Rust code
        --no-color              Disable color detection
    -h, --help                  Prints help information
```

Reply:
I've pushed a verbose approach that works. I'll see if I can get a subcommand to work tomorrow. Thanks for all your assistance so far!

Reply:
Please check the subcommand approach and let me know if it's not working for you. I would like to improve code/docs so users don't have to make verbose hacks.

Reply:
On an unrelated note, https://github.com/gnzlbg/cargo-asm is a super useful tool, but the author is unreachable, so no new releases are being made. I would be very happy to see a successor to it.

Reply:
You can already use it if you are on linux/macos and x86. Some things are not quite there yet but dumping asm for anything should just work.

`cargo install --git https://github.com/pacak/cargo-show-asm`

Reply:
I have spent half an hour messing around with types, and failed to get something that compiles and makes it so all of these are accepted:
```
supply-chain crates -d
crates -d
supply-chain update
update
```
with the help text pretending that the `supply-chain` argument doesn't exist.

Perhaps there is some clever way to do that, but I have failed to find it.

Reply:
Are there any good reasons to support versions both with and without `supply-chain` prefix? I'll add a few examples, might need to write some code for that if it's missing.

Reply:
It's not essential, but it does simplify contributions significantly, since anyone can just use `cargo run` instead of `cargo run -- supply-chain`.

I suppose showing a very helpful error message in absence of the `supply-chain` argument would also work.

Reply:
Okay, so that's mostly for you or anyone else who works on cargo-supply-chain repo directly. Makes sense. Will make something.

Reply:
published 0.3.2 with `cargo_helper` and some related examples